Fix streaming thinking tags split across multiple chunks

dsfaccini opened the pull request:
Fixes #3007

When streaming responses from models like Gemini via LiteLLM, thinking tags can be split across multiple chunks (e.g., chunk 1: `"<thi"`, chunk 2: `"nk>content</think>"`). The existing implementation only detected complete tags that arrived as standalone chunks, causing split tags to be treated as regular text instead of being extracted into `ThinkingPart`.

## Changes:
- Added buffering mechanism to `ModelResponsePartsManager` to accumulate content when it might be part of a split tag
- Refactored `handle_text_delta()` to detect complete tags across chunk boundaries while maintaining backward compatibility
- Added comprehensive tests covering 2-chunk splits, 3+ chunk splits, false positives, and interleaved content scenarios (thinking tags and text mixed/interleaved together)
- Models without thinking tags (e.g., Anthropic with native thinking support) are unaffected

## Edit: what does this PR do

### Constraint 1:
- `_parts_manager.py::get_parts()` returns `-> Generator[ModelResponseStreamEvent, None, None]` 
  - instead of `-> ModelResponseStreamEvent | None`
- this allows us to return multiple events instead of 0 or 1
- also has a nice side-effect of replacing multiple `if is None` checks with `for event in ...`

### Constraint 2:
- It doesn't change any existing test aside from adapting to the new return type of the `_parts_manager`
- I created a new test file for this specific case (split thinking tags)

### Constraint 3:
- The new functions in `_parts_manager.py` will buffer chunks when a chunk arrives that looks like it will be a `<think>` tag
- But only if the chunk starts with sth that looks like a think tag, like: `<thi`
- That means a chunk like `foo<thi` will not get buffered, it'll be emitted as a `TextPart`

## Cases it does (should) and doesn't cover

✅ **split thinking tags** or **full thinking tags**
- ✅ **Split but full**: `<thi` + `nk>thinking</th` + `ink>` -> `ThinkingPart("thinking")`
- ✅ **Split, missing closing tag**: `<thi` + `nk>th` + `inking` -> `ThinkingPart("thinking")`
- ✅ **Same as above in one chunk**: `<think>thinking` -> `ThinkingPart("thinking")`
  - there's an [**open question**](https://github.com/pydantic/pydantic-ai/pull/3206#discussion_r2483976551) of whether a `<think>` with content, like in the previous examples, should be emitted as a `TextPart` if the tag isn't closed by the end of the stream. The current behavior (linked in the caveat) emits the `ThinkingPart` as soon as the tag comes along, so it would be consistent to do the same here
- ✅ `<think>` -> gets buffered until there's content for it
- **edge cases**: for both partial (`<th`) and full (`<think>`) starting tags, if there's no content at the end of the stream, they'll be emitted as `TextPart`s
  - **caveat**: this differs from the current `main` behavior, which immediately emits a `PartStartEvent` with an empty thinking part (`ThinkingPart("")`) by looking at [_parts_manager.py](https://github.com/pydantic/pydantic-ai/blob/main/pydantic_ai_slim/pydantic_ai/_parts_manager.py#L204-L211)

❌ **thinking tags after a text part**
- ❌ `foo<th` -> thinking chunk needs to **start** with something that looks like a thinking part
- ❌ `foo<think>` -> the above applies even if the tag is complete
- ❌ `foo` + `<thi` + `ink>` + ... -> `TextPart("foo<think>...")` -> all text
  - <details><summary>checks if text parts have been created for this vendor id</summary>pydantic/pydantic-ai/pull/3206</details>

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -58,6 +58,8 @@ class ModelResponsePartsManager:
     """A list of parts (text or tool calls) that make up the current state of the model's response."""
     _vendor_id_to_part_index: dict[VendorId, int] = field(default_factory=dict, init=False)
     """Maps a vendor's "part" ID (if provided) to the index in `_parts` where that part resides."""
+    _tag_buffer: dict[VendorId, str] = field(default_factory=dict, init=False)
```
```suggestion
    _thinking_tag_buffer: dict[VendorId, str] = field(default_factory=dict, init=False)
```

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -99,6 +104,33 @@ def handle_text_delta(
+                ignore_leading_whitespace=ignore_leading_whitespace,
+            )
+
+        # Original logic for non-thinking-tag case
```
"Original" doesn't make sense to someone reading the code now :)

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -152,6 +170,127 @@ def handle_text_delta(
+        start_tag, end_tag = thinking_tags
+
+        # Combine any buffered content with the new content
+        buffered = self._tag_buffer.get(vendor_part_id, '') if vendor_part_id is not None else ''
```
What if `vendor_part_id` is `None`? Will none of this work anymore? Should we require one for this method to work?

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -152,6 +170,127 @@ def handle_text_delta(
+                self._tag_buffer.pop(vendor_part_id, None)
+
+                # Process any remaining content after the end tag
+                if after_end:
```
As @phemmer said on the other issue, we need to support `foo<think>bar</think>baz` and return 3 events

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -152,6 +170,127 @@ def handle_text_delta(
+                # Clear any state for this vendor_part_id and start thinking part
+                self._vendor_id_to_part_index.pop(vendor_part_id, None)
+                self._tag_buffer.pop(vendor_part_id, None)
+                thinking_event = self.handle_thinking_delta(vendor_part_id=vendor_part_id, content='')
```
If there's `after_start`, we shouldn't need `content=''`

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -152,6 +170,127 @@ def handle_text_delta(
+            return False
+        # Check if the tag starts with any suffix of the content
+        # E.g., for content="<thi" and tag="<think>", we check if "<think>" starts with "<thi"
+        for i in range(len(content)):
```
We don't need to look at the entire content, right, just the last `len(tag)` chars?

dsfaccini replied:
This was replaced by the function `_parts_manager.py::_could_be_tag_start()`

```python
    def _could_be_tag_start(self, content: str, tag: str) -> bool:
        """Check if content could be the start of a tag."""
        # Defensive check for content that's already complete or longer than tag
        # This occurs when buffered content + new chunk exceeds tag length
        # Example: buffer='<think' + new='<' = '<think<' (7 chars) >= '<think>' (7 chars)
        if len(content) >= len(tag):
            return False
        return tag.startswith(content)
```

DouweM reviewed (commented):
@dsfaccini Thanks for working on this David. I think we need to test every plausible combination of strings and make sure we never lose text or events.

dsfaccini commented:
> @dsfaccini Thanks for working on this David. I think we need to test every plausible combination of strings and make sure we never lose text or events.

@DouweM thank you for taking the time to review it! I refactored the event handler into returning a generator and am taking into account your comments. And I agree with your comment about the XML case. Specially considering this is a very edge case, and it might even fix itself in the future (by producing chunks properly).

I have a question, is it weird that all checks passed even though there may be a lot of breaking stuff in the PR? Does it mean that we should add tests to cover these cases or is there something I may be misunderstanding about the testing/CI process?

DouweM commented:
I refactored the event handler into returning a generator and am taking into account your comments.

@dsfaccini Did you mean to have pushed already?

> 
> I have a question, is it weird that all checks passed even though there may be a lot of breaking stuff in the PR? Does it mean that we should add tests to cover these cases or is there something I may be misunderstanding about the testing/CI process?

It doesn't break any existing tests since there are none that currently hit the `<think>`-tag related edge cases we're identifying. So we should focus on making the new test suite very exhaustive with all the edge cases we can think of.

dsfaccini commented:
> @dsfaccini Did you mean to have pushed already?
> 
> > I have a question, is it weird that all checks passed even though there may be a lot of breaking stuff in the PR? Does it mean that we should add tests to cover these cases or is there something I may be misunderstanding about the testing/CI process?
> 
> It doesn't break any existing tests since there are none that currently hit the `<think>`-tag related edge cases we're identifying. So we should focus on making the new test suite very exhaustive with all the edge cases we can think of.

No I didn't mean to push anything yet, I wanted to clear up that question to make sure I was not misunderstanding something about the test suite. Also I didn't wanna push anything else before clearing the XML "case".

> Maybe we should only do this at the start of a response, not allowing <think> portions in the middle of text output. And/or leave this off by default and require a ModelProfile setting to opt into it.

For now I'll write new tests to cover the edge cases you've pointed out, excluding the possibility (quoted above) for `<think>` tags in the middle of text.

dsfaccini commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -152,6 +170,127 @@ def handle_text_delta(
+                else:
+                    # No content after start tag
+                    return text_event if text_event is not None else thinking_event
+            elif self._could_be_tag_start(combined_content, start_tag):
```
as per the constraints this would become a `TextPart` with content `foo<thi`

dsfaccini commented on tests/test_parts_manager.py:
```diff
@@ -162,6 +162,203 @@ def test_handle_text_deltas_with_think_tags():
+
+    # Build up: "text <thi"
+    event = manager.handle_text_delta(vendor_part_id='content', content='text <thi', thinking_tags=thinking_tags)
+    assert event is None
```
this remains a valid concern: if the last chunk happens to be `<thi`, it will be lost (because it will be buffered but the parts manager won't be called again)

I'll add a test for this and remediate

DouweM commented on tests/test_parts_manager_split_tags.py:
```diff
@@ -0,0 +1,573 @@
+    events = list(manager.handle_text_delta(vendor_part_id='content', content='nk>', thinking_tags=thinking_tags))
+    assert len(events) == 1
+    assert events[0] == snapshot(
+        PartStartEvent(index=0, part=ThinkingPart(content='', part_kind='thinking'), event_kind='part_start')
```
Since we're doing more buffering now, could we make it so that we don't yield a `ThinkingPart` with `content=''` and only yield once we see actual text? I did it like this before because we were checking the most recent part to see if it was a `ThinkingPart`, to determine whether we're in "thinking mode", but we can now keep track of that more directly

DouweM commented on tests/test_parts_manager_split_tags.py:
```diff
@@ -0,0 +1,573 @@
+
+    # Complete start tag with vendor_part_id=None goes through simple path
+    # This covers lines 161-164 in _handle_text_delta_simple
+    events = list(manager.handle_text_delta(vendor_part_id=None, content='<think>', thinking_tags=thinking_tags))
```
If the model only ever outputs `<think>`, without any text following it, I think that should be a `<think>` text part

dsfaccini commented on tests/models/test_model_test.py:
```diff
@@ -342,3 +342,24 @@ def test_different_content_input(content: AudioUrl | VideoUrl | ImageUrl | Binar
+    Note: TestModel doesn't pass thinking_tags during streaming, so this doesn't actually
+    test buffering behavior - it just verifies that calling get() works correctly.
+    The actual buffering logic is thoroughly tested in test_parts_manager_split_tags.py,
+    and normal streaming is tested extensively in test_streaming.py.
```
I believe there was a couple test cases just for coverage, I can remove this one and add a pragma instead

dsfaccini commented on tests/test_parts_manager_split_tags.py:
```diff
@@ -0,0 +1,573 @@
+    # Add both empty and non-empty content to test the branch where buffered_content is falsy
+    # This ensures the loop continues after skipping the empty content
+    manager._thinking_tag_buffer['id1'] = ''  # Will be skipped  # pyright: ignore[reportPrivateUsage]
+    manager._thinking_tag_buffer['id2'] = 'content'  # Will be flushed  # pyright: ignore[reportPrivateUsage]
```
coverage, will remove and add back the pragmas

DouweM replied:
We should be able to get through every code path through public API usage right?

dsfaccini commented on tests/test_parts_manager.py:
```diff
@@ -546,7 +561,7 @@ def test_handle_thinking_delta_wrong_part_type():
 
     # Add a text part first
-    manager.handle_text_delta(vendor_part_id='text', content='hello')
+    list(manager.handle_text_delta(vendor_part_id='text', content='hello'))
```
cause the `handle_text_delta` now produces a generator, if we don't iterate over it it doesn't get evaluated at all

dsfaccini commented on tests/test_parts_manager_split_tags.py:
```diff
@@ -0,0 +1,573 @@
+    )
+
+    # Chunk 2: "<thi" - starts at position 0 of THIS chunk, buffer it
+    events = list(manager.handle_text_delta(vendor_part_id='content', content='<thi', thinking_tags=thinking_tags))
```
yeah, you're right, I didn't understand very well how chunks pass through the parts manager, looking at this now I think I what I need to specify is: **if a text part has been created for a vendor id, disregard any thinking tags that later appear for that vendor id**

so I'll use the `_vendor_id_to_part_index` to track if text parts have been created for a vendor id and "disable" thinking tag recognition in those cases

dsfaccini commented on tests/test_parts_manager_split_tags.py:
I delayed this bc it's easier for me to see the diff on my IDE

dsfaccini commented on pydantic_ai_slim/pydantic_ai/models/__init__.py:
```diff
@@ -553,6 +553,10 @@ async def _get_event_iterator(self) -> AsyncIterator[ModelResponseStreamEvent]:
     def get(self) -> ModelResponse:
         """Build a [`ModelResponse`][pydantic_ai.messages.ModelResponse] from the data received from the stream so far."""
+        # Flush any buffered content before building response
+        for _ in self._parts_manager.finalize():
```
added it to `__aiter__`. I realized calling `get` mid-stream would empty the buffer, messing up split-tag handling, so I'm addressing that in the next commit by cloning the parts manager in the `get` method.

DouweM commented on tests/test_parts_manager_split_tags.py:
```diff
@@ -0,0 +1,349 @@
+
+
+def stream_text_deltas(
+    chunks: list[str],
```
I'd also like to test a scenario like `['<thi', 'nk>foo', 'bar</th', ToolCallPart()]`, in which case the thinking part would presumably have content `foobar</th`. So maybe we can let this also take parts that fed to other `handle_...` methods?

dsfaccini replied:
currently, the closing `</th` would get stuck in the closing buffer of the `ThinkingPart`; the current implementation tolerates "orphaned" partial closing tags (like `</th`). We could flush them in the `final_flush` method but that would only flush them at the end of the stream.

Otherwise we could have an "append" hook on `self._parts` that flushes them whenever a new part is added and the previous part was a `ThinkingPart` with an orphaned buffer. The caveat here is that a `ThinkingPart` with a buffer isn't necessarily the latest part, given that we only buffer content for parts with a vendor id. So the danger would be that we preemptively flush it and then we get the remaining chunk.

So currently the alternatives would be:
- tolerate the orphaned buffers (🥇)
- flush them at the end of the stream (🥈)
- preemptively flush them when a new part comes along (🥉)

DouweM commented on pydantic_ai_slim/pydantic_ai/models/__init__.py:
```diff
@@ -580,6 +581,16 @@ def part_end_event(next_part: ModelResponsePart | None = None) -> PartEndEvent |
+                            if end_event:
+                                yield end_event
+                        last_start_event = finalize_event
+                    yield finalize_event
```
Should we set `finalize_event.previous_part_kind` like we do above? Could we reuse that same logic instead of duplicating it?

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -47,6 +47,75 @@
+    combined = buffered + content
+    segments: list[tuple[str, str]] = []
+    current_thinking_state = in_thinking
+    remaining = combined
```
We don't use `combined` anymore, so we can have just one variable right?

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -47,6 +47,75 @@
+            else:
+                # Check for partial end tag at end of remaining content
+                for i in range(len(remaining)):
+                    suffix = remaining[i:]
```
We're creating a lot of suffixes that we'll immediate discard. I think we should be able to iterate only over the actual suffixes but using `remaining[-N:]` to look at the last N chars, and then fill for N use `1` through `len(end_tag)` (with a check for the length of remaining)

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -67,6 +142,80 @@ def get_parts(self) -> list[ModelResponsePart]:
+                    self._started_part_indices.add(part_index)
+
+        # flush any remaining buffered content
+        for vendor_part_id, buffered_content in list(self._thinking_tag_buffer.items()):
```
`list` shouldn't be needed here

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -75,82 +224,254 @@ def handle_text_delta(
+                part_index = len(self._parts) - 1
+                latest_part = self._parts[part_index]
+                if isinstance(latest_part, ThinkingPart):
+                    yield from self.handle_thinking_delta(vendor_part_id=None, content=content)
```
Why do we need to do this? If there's no `vendor_part_id` (and we only do thinking part parsing when there is a `vendor_part_id`), wouldn't this be a new text part?

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -75,82 +224,254 @@ def handle_text_delta(
         Raises:
             UnexpectedModelBehavior: If attempting to apply text content to a part that is not a TextPart.
         """
+        if thinking_tags and vendor_part_id is not None:
```
I'm not sure why this should depend on their being a `vendor_part_id`. Couldn't we store a buffer for `None` as well?

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -75,82 +224,254 @@ def handle_text_delta(
+            if segment_type == 'text':
+                # Skip whitespace-only text before a thinking tag when ignore_leading_whitespace=True
+                skip_whitespace_before_tag = (
+                    ignore_leading_whitespace
```
Maybe it would make sense to move this whitespace logic to the chunking function

dsfaccini commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -67,20 +118,49 @@ def get_parts(self) -> list[ModelResponsePart]:
-        When `vendor_part_id` is None, the latest part is updated if it exists and is a TextPart;
-        otherwise, a new TextPart is created. When a non-None ID is specified, the TextPart corresponding
-        to that vendor ID is either created or updated.
+        This function also handles what we'll call "loose thinking", which is the generation of
```
I went Shakespeare on this docstring, let me know if it's too verbose. I thought it'd be good to document the limits of handling this in the code, could make it easier in the future to debug or adapt.

dsfaccini commented on tests/test_parts_manager_split_tags.py:
```diff
@@ -0,0 +1,349 @@
+    )
+    assert len(parts) == 2
+    assert parts[0] == snapshot(ThinkingPart(content='reasoning content', part_kind='thinking'))
+    assert parts[1] == snapshot(TextPart(content='after', part_kind='text'))
```
the cases are now clearer in [tests/test_parts_manager_thinking_tags.py](https://github.com/pydantic/pydantic-ai/pull/3206/files#diff-fef6a48941d9ebfd390fa8ce13628bd10af147abac4a0fc7432edf4e2f911fbf) but I didn't use snapshots, let me know if I should add snapshots.

dsfaccini commented:
Hey @DouweM , I went back to basics with this one. Can't say that I simplified the parts manager but I hope I at least made it easier to understand.

One thing that makes the current (main branch's) `_parts_manager::handle_text_delta` complicated is that there isn't a clean-cut exclusion of conditions because multiple conditions are checked at the same time, so the typechecker later can't be sure about the state of either variable.

So a big part of this PR is making those states very clear in the comments and adding a lot of info about handling 'loose' thinking tags in the docstring. Anyway, I added some comments to draw your attention to some specific parts, but I'll leave a summar here as well:

Summary of changes:
- decoupled `handle_text_delta` from `handle_thinking_delta`
- added an explanation to the docstring about 'loose thinking' to differentiate it from `handle_thinking_delta`
- changed the description of the `_vendor_id_to_part_index`
- added two sub-functions: `_handle_text_with_thinking_opening` and `_handle_text_with_thinking_closing`
- each handles its own case, `...opening` duplicates a bit of logic of `...closing` but I was worried abstracting that logic will make the whole thing less readable
- documented the limits of the handle and a couple of edge cases
- added the `tests/test_parts_manager_thinking_tags.py` and replaced the function-by-function structure with a parametrized version (should be easier to review each case by itself!)

DouweM commented on pydantic_ai_slim/pydantic_ai/messages.py:
```diff
@@ -1009,6 +1014,9 @@ class ThinkingPart:
     part_kind: Literal['thinking'] = 'thinking'
     """Part type identifier, this is available on all parts as a discriminator."""
 
+    closing_tag_buffer: Annotated[str, pydantic.Field(exclude=True)] = field(compare=False, default='', repr=False)
```
Same as up, we'll need to store this on the parts manager somehow

DouweM commented on pydantic_ai_slim/pydantic_ai/models/__init__.py:
```diff
@@ -615,7 +616,17 @@ def part_end_event(next_part: ModelResponsePart | None = None) -> PartEndEvent |
+                ) -> AsyncIterator[ModelResponseStreamEvent]:
+                    async for event in iter1:
+                        yield event
+                    for event in iter2:  # pragma: no cover - idk why this isn't covered
```
we'll definitely want to make sure this is covered :)

dsfaccini replied:
this line is definitely running, the tests include `final_flush()` behavior and `final_flush()` in only called here, I just don't know why coverage isn't picking it up... the current CI coverage failure seems to agree that it's being covered but if I remove it and coverage fails in the next run I don't know what I'll do 🫣

DouweM commented on pydantic_ai_slim/pydantic_ai/models/__init__.py:
```diff
@@ -615,7 +616,17 @@ def part_end_event(next_part: ModelResponsePart | None = None) -> PartEndEvent |
+                    for event in iter2:  # pragma: no cover - idk why this isn't covered
+                        yield event
+
+                async for event in chain_async_and_sync_iters(  # pragma: no cover - related to above
```
and this

DouweM commented on tests/test_parts_manager_thinking_tags.py:
```diff
@@ -0,0 +1,549 @@
+    chunks: list[str]
+    expected_parts: list[ModelResponsePart]  # [TextPart|ThinkingPart('final content')]
+    expected_events: Sequence[ModelResponseStreamEvent]
+    expected_events_before_flushing: Sequence[ModelResponseStreamEvent] | Literal['same-as-expected-events'] = (
```
This may be clearer if we could have `expected_flushed_events` default to `[]`

DouweM commented on tests/test_parts_manager_thinking_tags.py:
```diff
@@ -0,0 +1,549 @@
+        expected_events=[
+            PartStartEvent(index=0, part=ThinkingPart('content')),
+        ],
+        leftover_closing_bufffer=['</th'],
```
I think we should flush this as a thinking delta in cases like this

DouweM commented on tests/test_parts_manager_thinking_tags.py:
```diff
@@ -0,0 +1,549 @@
+    ),
+    Case(
+        name='new_part_ignore_whitespace_mixed_with_partial_opening',
+        chunks=['  <thi'],
```
We should also test whitespace followed by an actual think tag

DouweM commented on tests/test_parts_manager_thinking_tags.py:
```diff
@@ -0,0 +1,549 @@
+    Case(
+        name='existing_text_stutter_via_append_non_empty_content',
+        chunks=['hello', '<thi', '<think>content</think>'],
+        expected_parts=[TextPart('hello<thi'), ThinkingPart('content')],
```
Same as above

DouweM commented on tests/test_parts_manager_thinking_tags.py:
```diff
@@ -0,0 +1,549 @@
+FAKE_CLOSING_CASES: list[Case] = [
+    Case(
+        name='existing_thinking_fake_closing_added_to_thinking',
+        chunks=['<think>content', '</fake>'],
```
This may be more interesting to test as `'</', 'fake>'`, as it starts looking like a closing tag but then isn't

DouweM commented on tests/test_parts_manager_thinking_tags.py:
```diff
@@ -0,0 +1,549 @@
+
+
+@pytest.mark.parametrize('case', ALL_CASES, ids=lambda c: c.name)
+def test_thinking_parts_parametrized(case: Case) -> None:
```
This is awesome, nice work

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -67,20 +118,49 @@ def get_parts(self) -> list[ModelResponsePart]:
+        This function also handles what we'll call "loose thinking", which is the generation of
+        ThinkingParts via explicit thinking tags embedded in the text content.
+        Activating loose thinking requires:
+        - `thinking_tags` to be provided, which is a tuple of (opening_tag, closing_tag)
```
Please wrap all code in backticks, i.e. `(opening_tag, closing_tag)` here, `ThinkingPart` above, `vendor_part_id` below, etc

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -99,58 +179,385 @@ def handle_text_delta(
             # Otherwise, attempt to look up an existing TextPart by vendor_part_id
             part_index = self._vendor_id_to_part_index.get(vendor_part_id)
             if part_index is not None:
                 existing_part = self._parts[part_index]
```
I wonder if a helper method like `_existing_part_by_vendor_id()` could be helpful

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -99,58 +179,385 @@ def handle_text_delta(
             UnexpectedModelBehavior: If attempting to apply text content to a part that is not a TextPart.
         """
-        existing_text_part_and_index: tuple[TextPart, int] | None = None
+        potential_part: _ExistingPart[TextPart] | _ExistingPart[ThinkingPart] | None = None
```
Can we call this `existing_part`?

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -99,58 +179,381 @@ def handle_text_delta(
+                # EC2: delayed thinking - we don't emit an event until there's content after the tag
+                yield from _buffer_thinking()  # RO1
+            elif combined_buffer.startswith(opening_tag):
+                # TODO this whole elif is very close to a duplicate of `_handle_text_with_thinking_closing`,
```
Same file is fine, embedded thinking is the hardest part of handling these events, not unreasonably so

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -99,58 +179,385 @@ def handle_text_delta(
+            elif combined_buffer.startswith(opening_tag):
+                # TODO this whole elif is very close to a duplicate of `_handle_text_with_thinking_closing`,
+                #   but we can't delegate because we're generating different events (starting ThinkingPart vs updating it)
+                #   and there's no easy abstraction that comes to mind, so I'll leave it as is for now.
```
Once we have the whole test suite complete, and we've addressed some of the suggestions I made earlier, I wonder if we can have Claude Opus do a full refactor while maintaining the tests green

DouweM replied:
One approach worth considering would be to not try to go directly from text chunks to Parts and Events, but to first go from chunks to an intermediate representation (similar to what you had before I think) with concepts like "text", "partial_start_tag", "thinking", "partial_end_tag", and then to have separate logic that converts that representation into Parts and Events

dsfaccini replied:
ftr the current logic also doesn't emit events for partial tags, but I'll definitely migrate to IRs, that seems like the next easiest option in lieu of adding buffer attributes to the parts

your `expected_next` idea is excellent! I'm thinking that'll make lots of logic clearer

there are two points in this comment I think, 
1. is about the buffering and event emission logic
2. is about the IRs

### About the logic
- the incomplete start tag is kept in the buffer of the latest `TextPart`, if there's not latest `TextPart` it **will** be created, but no event is emitted for it, it just keeps a placeholder in `self._parts`
- if the opening tag completes without content, that `TextPart` remains with a full buffer, waiting for content, but no event is emitted still
- if content comes along that `TextPart` is replaced in its index with a `ThinkingPart` and the `PartStartEvent` is finally emitted
- otherwise, if content comes along that looks like it's generating an empty `ThinkingPart` (i.e. `TextPart(buffer='<think>', content='</th')`) the current logic will emit a `PartStartEvent(part=TextPart(content='<think></th')`, i.e. it will give up on waiting for valid content.
  - if the content that comes along effectively closes what would've been the `ThinkingPart`, the logic currently emits an event for a `TextPart` or `TextPartDelta` with `content='<think></think>'`
  - `# 1.a. '<think></think>more content'` -> `yield from handle_invalid_opening_tag()  # RO2`
  - `<th` -> no event, `TextPart(buffer='<th')` -> potentially thinking
  - `ink>` -> no event, `TextPart(buffer='<think>')` -> potentially thinking
  - `</th` -> `PartStartEvent` with `TextPart(buffer='', content='<think></th')` -> give up
  - because we're giving up early, the only case currently left uncovered is:
  - `ankyou very much</think>` -> ThinkingPart(content='</thankyou very much')` -> false negative!
- the IR solution may help dealing with this last case!


### About the IRs
I'll definitely migrate the PR to using `PartialStartTag` and `PartialEndTag` in a separate tracker dict, that seems like the next easiest solution.

dsfaccini replied:
I wrote this huge comment yesterday but never submitted it, my bad.

DouweM commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -99,58 +179,381 @@ def handle_text_delta(
-
-        if existing_text_part_and_index is None:
+        if potential_part is None:
             # This is a workaround for models that emit `<think>\n</think>\n\n` or an empty text part ahead of tool calls (e.g. Ollama + Qwen3),
```
I suggest checking the commit/PR that came from and making sure there's a test that still passes

DouweM reviewed (changes requested):
@dsfaccini I appreciate your perseverance 😄

dsfaccini commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -99,58 +179,385 @@ def handle_text_delta(
                 if isinstance(latest_part, TextPart):
-                    existing_text_part_and_index = latest_part, part_index
+                    potential_part = _ExistingPart(part=latest_part, index=part_index, found_by='latest_part')
+                    # ✅ vendor_part_id and ✅ potential_part is a TextPart
```
the colors helped me lol, I thought they were useful, will remove in the next push

dsfaccini commented on tests/test_parts_manager.py:
```diff
@@ -28,13 +27,17 @@ def test_handle_text_deltas(vendor_part_id: str | None):
+    events = manager.handle_text_delta(vendor_part_id=vendor_part_id, content='hello ')
+    assert len(events) == 1
+    event = events[0]
     assert event == snapshot(
```
yeah I just modified the existing tests to adapt to `Generator` behavior but I'll replace all the assertions with snapshots as well 👌

dsfaccini commented on tests/test_parts_manager.py:
```diff
@@ -529,12 +463,12 @@ def test_handle_thinking_delta_no_vendor_id_with_existing_thinking_part():
 
     # Now add another thinking delta with no vendor_part_id - should update the latest thinking part
-    event = manager.handle_thinking_delta(vendor_part_id=None, content=' more', signature=None)
+    event = next(manager.handle_thinking_delta(vendor_part_id=None, content=' more', signature=None))
```
this test tests a `ThinkingPart` that gets created with a valid `vendor_part_id` and is later updated with a `vendor_part_id=None`

dsfaccini commented on tests/test_parts_manager_thinking_tags.py:
```diff
@@ -0,0 +1,549 @@
+        expected_parts=[ThinkingPart('content'), TextPart('more')],
+        expected_events=[
+            PartStartEvent(index=0, part=ThinkingPart('content')),
+            PartStartEvent(index=1, part=TextPart('more')),
```
I'm just following the previous logic here, the previous tests weren't testing the case when a new `ThinkingPart` is created and its vendor id **is** None 

following the code of the main branch result of a embedded thinking without a vendor id would be to produce a `ThinkingPart` and forgetting about it completely because:
1. it's not tracked in `_vendor_id_to_part_index` because we always check `vendor id is not None`
2. and the `handle_text_delta` logic completely ignores the latest part if it's not a `TextPart`, it only handles a `ThinkingPart` that it finds via vendor id.

So I could update `_append_and_track_new_part` to allow tracking of vendor ids that are `None` but this would break with previous behavior

This test was testing an existing `ThinkingPart` that gets updated (without a vendor id)
https://github.com/pydantic/pydantic-ai/pull/3206/files#diff-798decb664333b771fca30e50ec8c50474132dfb5e8e843d0542ea4629782289L532-R471

This one was testing a new `ThinkingPart` (with a vendor id)
https://github.com/pydantic/pydantic-ai/pull/3206/files#diff-798decb664333b771fca30e50ec8c50474132dfb5e8e843d0542ea4629782289L100-L108

dsfaccini replied:
I was able to set this up, the latest `test_parts_manager_thinking_tags.py` runs the test with both a `vendor_part_id='content'` and `vendor_part_id=None`

dsfaccini commented on pydantic_ai_slim/pydantic_ai/models/__init__.py:
```diff
@@ -615,7 +616,17 @@ def part_end_event(next_part: ModelResponsePart | None = None) -> PartEndEvent |
                     )
 
-                async for event in iterator:
+                async def chain_async_and_sync_iters(
```
ok this is now implemented in the last version

phemmer commented on pydantic_ai_slim/pydantic_ai/_parts_manager.py:
```diff
@@ -152,6 +170,127 @@ def handle_text_delta(
+                return self.handle_thinking_delta(vendor_part_id=vendor_part_id, content=combined_content)
+        else:
+            # Not in thinking mode, look for start tag
+            if start_tag in combined_content:
```
So FYI, I just ran across some new behavior by Gemini. In the very middle of the response it was sending, it injects:
```
*Self-correction: The above implementation is not quite right. The handlers should not directly add to the UI. They should emit JComponents. The following steps will build this out correctly.*
```
and later
```
*This is getting too complex and I'm re-implementing Flow operators badly. The user's initial feedback was about a blocking finalize. The job.join() approach is much simpler and more direct. I will revert to that simpler, more robust model and abandon this re-implementation.*
```

These are clearly intended to be thoughts. They're currently exposed as plain text in the response itself, but this indicates that Google is experimenting with putting thoughts in the middle of responses. This is obviously just an experiment, as it lacked any sort of tags and just used `*` instead. But it shows that we may be dealing with this soon. I don't think it should block this PR, as this PR is an obvious improvement in the right direction. Just wanted to throw this out as a "heads up, it's coming".

DouweM commented:
@dsfaccini Can you please merge main and fix the conflicts before I review again?

dsfaccini commented:
hey @xcpky thank you for expressing your interest in this PR, there have been so many changes since then that it's not gonna be trivial to resolve the merge conflicts but I'll put some effort into it next week

dsfaccini commented on pydantic_ai_slim/pydantic_ai/_thinking_part.py:
```diff
@@ -29,3 +29,8 @@ def split_content_into_text_and_thinking(content: str, thinking_tags: tuple[str,
+
+# NOTE: this utility is used by models/: `groq`, `huggingface`, `openai`, `outlines` and `tests/test_thinking_part.py`
+# not sure if it could be replaced by the new handling in the `_parts_manager.py` but it's worth taking a closer look.
+# if that's the case we could use this file to partly isolate the embedded thinking handling logic and declutter the parts manager.
```
leaving this in for discussion
